File: app/src/main/java/com/example/all_sensors/presentation/TrackerManager.java
package com.example.all_sensors.presentation;

import android.content.Context;

import com.samsung.android.service.health.tracking.ConnectionListener;
import com.samsung.android.service.health.tracking.HealthTrackingService;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.HealthTrackerType;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Envuelve HealthTrackingService y cada HealthTracker.
 * Ahora recibe un ConnectionListener y solo arranca trackers tras onConnectionSuccess().
 */
public class TrackerManager {
    private final HealthTrackingService sdk;
    private final Map<HealthTrackerType, HealthTracker> cache = new HashMap<>();

    /***** Métodos package-private para que el ConnectionListener pueda avisar *****/
    private boolean isConnected = false;

    public void markConnected() {
        isConnected = true;
    }

    public void markDisconnected() {
        isConnected = false;
    }
    private void ensureConnected() {
        if (!isConnected) {
            throw new IllegalStateException("HealthTrackingService aún no conectado");
        }
    }
    public boolean isConnected() {
        return isConnected;
    }


    /**
     * @param context  contexto
     * @param connListener  tu listener que recibe onConnectionSuccess/Failed
     */
    public TrackerManager(Context context, ConnectionListener connListener) {
        // Pasamos siempre un listener no-nulo
        sdk = new HealthTrackingService(connListener, context);
        sdk.connectService();
    }

    /** Devuelve (y cachea) el HealthTracker */
    private HealthTracker getTracker(HealthTrackerType type) {
        ensureConnected();
        return cache.computeIfAbsent(type, t -> sdk.getHealthTracker(t));
    }

    /** Arranca un tracker continuo */
    public void startContinuous(HealthTrackerType type,
                                HealthTracker.TrackerEventListener listener) {
        HealthTracker tr = getTracker(type);
        tr.setEventListener(listener);
    }

    /** Para un tracker continuo */
    public void stopContinuous(HealthTrackerType type) {
        HealthTracker tr = cache.get(type);
        if (tr != null) tr.unsetEventListener();
    }

    public void stopAllContinuous() {
        for (HealthTracker tr : cache.values()) {
            tr.unsetEventListener();
        }
    }

    /** On-demand bloqueante (igual que antes) */
    public synchronized void runOnDemand(HealthTrackerType type,
                                         HealthTracker.TrackerEventListener listener,
                                         long timeoutMs) throws InterruptedException {
        HealthTracker tr = getTracker(type);
        CountDownLatch latch = new CountDownLatch(1);
        HealthTracker.TrackerEventListener wrapped = new HealthTracker.TrackerEventListener() {
            @Override
            public void onDataReceived(java.util.List<com.samsung.android.service.health.tracking.data.DataPoint> dataPoints) {
                listener.onDataReceived(dataPoints);
                latch.countDown();
            }
            @Override public void onFlushCompleted() { listener.onFlushCompleted(); }
            @Override public void onError(com.samsung.android.service.health.tracking.HealthTracker.TrackerError err) {
                listener.onError(err);
                latch.countDown();
            }
        };
        tr.setEventListener(wrapped);
        latch.await(timeoutMs, TimeUnit.MILLISECONDS);
        tr.unsetEventListener();
    }

    public void shutdown() {
        stopAllContinuous();
        sdk.disconnectService();
        cache.clear();
    }
}

File: app/src/main/java/com/signals/smartwatch/sensors/samsung/TrackerManager.kt
package com.signals.smartwatch.sensors.samsung

import android.content.Context
import com.samsung.android.service.health.tracking.ConnectionListener
import com.samsung.android.service.health.tracking.HealthTracker
import com.samsung.android.service.health.tracking.HealthTrackingService
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.HealthTrackerType
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeoutOrNull

class TrackerManager(context: Context, connListener: ConnectionListener) {

    private val sdk = HealthTrackingService(connListener, context).also { it.connectService() }
    private val cache = mutableMapOf<HealthTrackerType, HealthTracker>()

    @Volatile private var connected = false

    fun markConnected()    { connected = true }
    fun markDisconnected() { connected = false }
    fun isConnected()      = connected

    private fun ensureConnected() {
        check(connected) { "HealthTrackingService aún no conectado" }
    }

    private fun getTracker(type: HealthTrackerType): HealthTracker {
        ensureConnected()
        return synchronized(cache) { cache.getOrPut(type) { sdk.getHealthTracker(type) } }
    }

    fun startContinuous(type: HealthTrackerType, listener: HealthTracker.TrackerEventListener) {
        getTracker(type).setEventListener(listener)
    }

    fun stopContinuous(type: HealthTrackerType) {
        cache[type]?.unsetEventListener()
    }

    fun stopAllContinuous() {
        cache.values.forEach { it.unsetEventListener() }
    }

    /**
     * Mantiene el listener activo durante [timeoutMs] para recibir todas las muestras
     * de la ventana. Termina antes solo si el SDK reporta error.
     */
    suspend fun runOnDemand(
        type: HealthTrackerType,
        listener: HealthTracker.TrackerEventListener,
        timeoutMs: Long
    ) {
        val tracker = synchronized(cache) { getTracker(type) }
        val errorSignal = CompletableDeferred<Unit>()

        val wrapped = object : HealthTracker.TrackerEventListener {
            override fun onDataReceived(dataPoints: List<DataPoint>) =
                listener.onDataReceived(dataPoints)
            override fun onFlushCompleted() = listener.onFlushCompleted()
            override fun onError(err: HealthTracker.TrackerError) {
                listener.onError(err)
                errorSignal.complete(Unit)
            }
        }
        tracker.setEventListener(wrapped)
        try {
            withTimeoutOrNull(timeoutMs) { errorSignal.await() }
        } finally {
            tracker.unsetEventListener()
        }
    }

    /** Inicia un nuevo intento de conexión tras una desconexión inesperada. */
    fun reconnect() {
        sdk.connectService()
    }

    fun shutdown() {
        stopAllContinuous()
        sdk.disconnectService()
        cache.clear()
    }
}
